bin/build-mb/src/build/licenses.clj

```diff
@@ -0,0 +1,201 @@
+(ns build.licenses
+  "Functionality to take a classpath and generate a file containing all libraries and their respective licenses.
+  Intended to be called from from the command line like:
+
+  clj -X mb.license/process :classpath a-classpath :output-filename a-filename
+
+  or from clojure code like:
+  ```
+  (license/process {:classpath classpath
+                    :backfill (edn/read-string (slurp (io/resource \"overrides.edn\")))
+                    :output-filename (.getAbsolutePath output-file)
+                    :report? false})
+  ```
+
+  Allows for optional backfill, either edn or a filename that is edn. Backfill should be of the form
+
+  {\"group\" {\"artifact\" \"license text\"}
+   \"group\" {\"artifact\" {:resource \"filename-of-license\"}}
+
+  :override/group {\"group\" \"license\"
+                   \"group\" {:resource \"filename-of-license\"}}
+  }
+
+  At the moment assumes a leiningen classpath which is primarily jars and your own source paths, so only detects
+  license information from jars. In the future a strategy and heuristics could be determined for when source is
+  available from local roots and git dependencies."
+  (:require [clojure.data.xml :as xml]
+            [clojure.edn :as edn]
+            [clojure.java.io :as io]
+            [clojure.string :as str])
+  (:import (java.util.jar JarFile JarFile$JarFileEntry)))
+
+(set! *warn-on-reflection* true)
+
+(def classpath-separator (System/getProperty "path.separator"))
+
+(defn jar-file? [filename]
+  (str/ends-with? filename "jar"))
+
+(defn jar-files [jar-filename]
+  (iterator-seq (.entries (JarFile. (io/file jar-filename)))))
```

Comment:
You could use NIO to get a specific file out of the JAR if you wanted (this is how we read the `metabase-plugin.yaml` file inside Metabase plugin JARs for example -- see `metabase.util.files`). 

```clj
(defn do-with-jar-file-input-stream [^String jar-path ^String file-path f]
  (let [jar-path (.getPath (java.nio.file.FileSystems/getDefault) jar-path (into-array String []))]
    (with-open [fs (java.nio.file.FileSystems/newFileSystem jar-path (ClassLoader/getSystemClassLoader))]
      (let [file-path (.getPath fs file-path (into-array String []))]
        (with-open [is (java.nio.file.Files/newInputStream file-path (into-array java.nio.file.OpenOption []))]
          (f is))))))

(defn slurp-jar-file [jar-path file-path]
  (do-with-jar-file-input-stream
   jar-path
   file-path
   (fn [^java.io.InputStream is]
     (slurp is))))

(slurp-jar-file "/home/cam/metabase/resources/modules/google.metabase-driver.jar" "metabase-plugin.yaml")
;; -> "info:\n  name: Metabase Google Drivers Shared Dependencies ..."
```

Not sure that's really any easier than just iterating over every file in the JAR but it would definitely be a lot faster for big ones.

Reply:
At least you could avoid dealing with `JarFile$JarFileEntry` nonsense. NIO abstracts out JARs into their own file system so you use the same functions to work with them as you do with regular files -- no need for a bunch of special-case classes like `JarFile`

Reply:
I iterate over every file in the jar searching for a pom. For random access, there's the handy 
```clojure
(defn get-entry [^JarFile jar ^String name]
  (.getEntry jar name))
```

Or if i'm looking for the first file at one of the license file locations,
```clojure
(defn license-from-jar
  [^JarFile jar]
  (when-let [license-entry (some (partial get-entry jar) license-file-names)]
    (with-open [is (.getInputStream jar license-entry)]
      (slurp is))))
```

Both seem pretty straightforward. I'm iterating over all jar entries when looking for the pom, and I only do this if there's no pom adjacent to the jar file, so not very often. If i wanted random access to the pom file, I'd need to know the path to it, which looks like the following:

```
("META-INF/maven/net.sf.cssbox/cssbox/pom.xml"
 "META-INF/maven/scout/scout/pom.xml"
 "META-INF/maven/commons-lang/commons-lang/pom.xml"
 "META-INF/maven/org.clojure/tools.analyzer.jvm/pom.xml")
```

If i were starting with coordinates that might be a good way to go but we are starting with a classpath so we'd have to parse jar filenames to get coordinates, which sounds like not so great.

As it stands I think this is fine for the time being.

In the future, especially if we switch over to deps.edn and clojure cli, we could leverage tools.deps.alpha as a library and compute a dep graph and proceed in this manner, starting from the dep tree. Or we could just extend the current way to look at folders on the classpath as well, since deps will either point us towards a jar or a gitlibs folder downloaded somewhere.